AutoHeuristic: Support ranking/pruning choices

AlnisM opened the pull request:
This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133611
* #133608
* #131714
* #131710
* __->__ #131705

AlnisM committed:
AutoHeuristic: Support ranking/pruning choices

[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

eellison commented on torchgen/_autoheuristic/train_decision.py:
```diff
@@ -102,8 +103,19 @@ def predict(self, model, df, feature_columns):
         leaf_ids = model.apply(df[feature_columns])
         return predictions, proba, leaf_ids
 
+    def ranking_num_choices(self):
+        # if the heuristic is used for ranking, this function returns the number
+        # of choices that the heuristic will return
+        return 5
```
I'm surprised this is hardcoded. Is this not configurable at all ?

AlnisM replied:
You can configure it by overriding the method https://github.com/pytorch/pytorch/pull/131714/files#diff-45bd9b9ea64020450836e3496f0da071f133e1065dc62fbbb141bd6f52d6d2b0R69. But I'll make this a command line option instead.

eellison reviewed (commented):
just one question about ranking num choices being hardcoded, also tests could b good

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM commented on torchgen/_autoheuristic/ah_tree.py:
```diff
@@ -0,0 +1,225 @@
+# mypy: ignore-errors
```
added some docstrings. I think some typehints could be better but I need to configure mypy to work with numpy types. happy to do this in a follow-up PR

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]

AlnisM committed:
Update on "AutoHeuristic: Support ranking/pruning choices"

This PR adds support in train_decision if one wants to learn a heuristic for ranking. The main idea is that the user has to provide a number of choices the heuristic should return. I added a way to prune the learned decision tree such that it always returns the number of choices provided by the user.




[ghstack-poisoned]
